README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ohsome: An R package to interact with the ohsome API for OpenStreetMap history data aggregation and extraction

<!-- badges: start -->

[![LICENSE](https://img.shields.io/github/license/GIScience/ohsome-r)](https://github.com/GIScience/ohsome-r/blob/main/LICENSE.md)
[![status:
active](https://github.com/GIScience/badges/raw/master/status/active.svg)](https://github.com/GIScience/badges#active)
![CRAN status](https://www.r-pkg.org/badges/version-ago/ohsome)
<!-- badges: end -->

This ohsome R package grants access to the power of the
<a href="https://api.ohsome.org" target="blank">ohsome API</a> from R.
ohsome lets you analyze the rich data source of the
<a href="https://www.openstreetmap.org/"
target="blank">OpenStreetMap</a> (OSM) history. It aims to leverage the
tools of the <a href="https://github.com/GIScience/oshdb"
target="blank">OpenStreetMap History Database</a> (OSHDB).

With ohsome, you can …

- Get **aggregated statistics** on the evolution of OpenStreetMap
  elements and specify your own temporal, spatial and/or thematic
  filters. The data aggregation endpoint allows you to access functions,
  e.g., to calculate the area of buildings or the length of streets at
  any given timestamp.

- Retrieve the **geometry** of the historical OpenStreetMap data, e.g.,
  to visualize the evolution of certain OpenStreetMap elements over
  time. You can get the geometries for specific points in time or all
  changes within a timespan (full-history).

## Installation

You can install ohsome from
<a href="https://CRAN.R-project.org/package=ohsome"
target="blank">CRAN</a> or from
<a href="https://github.com/GIScience/ohsome-r"
target="blank">GitHub</a>:

``` r
install.packages("ohsome")
remotes::install_github("GIScience/ohsome-r", ref = "dev-0.2.2.9000")
```

## Getting started

Upon attaching the ohsome package, a metadata request is sent to the
ohsome API. The package message provides some essential metadata
information, such as the current temporal extent of the underlying
OSHDB:

``` r
library(ohsome)
#> Data: © OpenStreetMap contributors https://ohsome.org/copyrights
#> ohsome API version: 1.10.1
#> Temporal extent: 2007-10-08T00:00:00Z to 2023-09-24T20:00Z
```

The metadata is stored in `ohsome_metadata`. You can print it to the
console to get more details.

You can create any ohsome API query using the generic `ohsome_query()`
function. It takes the endpoint path and any query parameters as inputs.
For information on all available endpoints with their parameters,
consult the
<a href="https://docs.ohsome.org/ohsome-api/stable/endpoints.html"
target="blank">ohsome API documentation</a> or print `ohsome_endpoints`
to the console.

However, this ohsome R package provides specific wrapper functions for
queries to all available endpoints.

### OSM elements

#### Aggregating OSM elements

The <a
href="https://docs.ohsome.org/ohsome-api/stable/endpoints.html#elements-aggregation"
target="blank">elements aggregation endpoints</a> of the ohsome API
allow querying for the aggregated amount, length, area or perimeter of
OpenStreetMap elements with given properties, within given boundaries
and at given points in time.

Let us create a query for the total amount of breweries on OSM in the
region of Franconia. The first argument to `ohsome_elements_count()` is
the `sf` object `franconia` that is included in the
<a href="https://r-spatial.github.io/mapview/"
target="blank">mapview</a> package and contains boundary polygons of the
37 districts of the region:

``` r
library(mapview)

q <- ohsome_elements_count(franconia, filter = "craft=brewery")
```

The resulting `ohsome_query` object can be sent to the ohsome API with
`ohsome_post()`. By default, `ohsome_post()` returns the parsed API
response. In this case, this is a simple `data.frame` of only one row.

``` r
ohsome_post(q, strict = FALSE)
#> Warning: The time parameter is not defined and defaults to the latest available timestamp within the underlying OSHDB.
#> You can use set_time() to set the time parameter.
#>             timestamp value
#> 1 2023-09-24 20:00:00   182
```

The `ohsome_query` object was created without an explicit `time`
parameter. When using functions to create element aggregation queries,
`time` defaults to the most recent available timestamp in the underlying
OSHDB.[^1]

Defining the `time` parameter unlocks the full power of ohsome API by
giving access to the OSM history. It requires one or more
<a href="https://docs.ohsome.org/ohsome-api/stable/time.html"
target="blank">ISO-8601 conform timestring(s)</a>. Here is how to create
a query for the number of breweries at the first of each month between
2010 and 2020:

``` r
ohsome_elements_count(franconia, filter = "craft=brewery", time = "2010/2020/P1M")
```

Alternatively, we can update the existing `ohsome_query` object `q` with
the `set_time()` function, pipe [^2] the modified query directly into
`ohsome_post()` and make a quick visualization with `ggplot2`:

``` r
library(ggplot2)

q |> 
    set_time("2010/2020/P1M") |>
    ohsome_post() |>
    ggplot(aes(x = timestamp, y = value)) +
    geom_line()
```

<img src="man/figures/README-pipe-1.svg" width="900" />

This is how to query the total number of breweries in all of Franconia.
But what if we want to aggregate the amount per district? The ohsome API
provides specific endpoints for different grouped calculations, such as
aggregation grouped by bounding geometry.

There are several ways to define a query for an aggregation grouped by
boundary:

The `set_endpoint`function is used to change or append the endpoint path
of an API request. In this case, we could append `groupBy/boundary` to
the existing query to the `elements/count` endpoint. The endpoint path
can either be given as a single string (`/groupBy/boundary`) or as a
character vector:
`set_endpoint(q, c("groupBy", "boundary"), append = TRUE)` [^3].

More comfortable, however, is the use of either the grouping argument
with an elements aggregation function (e.g. 
`ohsome_elements_count(grouping = "boundary)`) or of the
`set_grouping()` function to modify an existing query object:

``` r
library(dplyr)

franconia |> 
    mutate(id = NAME_ASCI) |>
    ohsome_elements_count(filter = "craft=brewery", time = "2021-06-01") |>
    set_grouping("boundary") |>
    ohsome_post()
#> Simple feature collection with 37 features and 3 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XY
#> Bounding box:  xmin: 8.975926 ymin: 48.8625 xmax: 12.27535 ymax: 50.56422
#> Geodetic CRS:  WGS 84
#> First 10 features:
#>             groupByBoundaryId value  timestamp                       geometry
#> 1   Bamberg, Kreisfreie Stadt     6 2021-06-01 MULTIPOLYGON (((10.92581 49...
#> 2  Bayreuth, Kreisfreie Stadt     6 2021-06-01 MULTIPOLYGON (((11.58157 49...
#> 3    Coburg, Kreisfreie Stadt     0 2021-06-01 MULTIPOLYGON (((10.95355 50...
#> 4       Hof, Kreisfreie Stadt     1 2021-06-01 MULTIPOLYGON (((11.93067 50...
#> 5          Bamberg, Landkreis    13 2021-06-01 MULTIPOLYGON (((10.87615 50...
#> 6         Bayreuth, Landkreis    13 2021-06-01 MULTIPOLYGON (((11.70656 50...
#> 7           Coburg, Landkreis     6 2021-06-01 MULTIPOLYGON (((10.88654 50...
#> 8                   Forchheim     8 2021-06-01 MULTIPOLYGON (((11.26376 49...
#> 9              Hof, Landkreis     4 2021-06-01 MULTIPOLYGON (((11.91989 50...
#> 10                    Kronach     1 2021-06-01 MULTIPOLYGON (((11.36979 50...
```

If you want your own identifiers for the geometries returned by ohsome,
your input `sf` object needs a column explicitly named `id`. You can use
`mutate()` or `rename()` from the
<a href="https://dplyr.tidyverse.org" target="blank">dplyr</a> package
to create such a column as in the example below.

By default, `ohsome_post()` returns an `sf` object whenever the ohsome
API is capable of delivering GeoJSON data. This is the case for elements
extraction queries as well as for aggregations grouped by boundaries.

Thus, you can easily create a choropleth map from the query results. In
addition, you can set the argument `return_value` to `density`. This
will modify the endpoint path of the query so that ohsome returns the
number of breweries per area instead of the absolute value:

``` r
franconia |> 
    mutate(id = NAME_ASCI) |>
    ohsome_elements_count(filter = "craft=brewery", return_value = "density") |>
    set_time("2021-06-01") |>
    set_grouping("boundary") |>
    ohsome_post() |>
    mapview(zcol = "value", layer.name = "Breweries per sqkm")
```

<img src="man/figures/README-density-1.png" width="900" />

#### Extracting OSM elements

The <a
href="https://docs.ohsome.org/ohsome-api/stable/endpoints.html#elements-extraction"
target="blank">elements extraction endpoints</a> of the ohsome API allow
obtaining geometries, bounding boxes or centroids of OSM elements with
given properties, within given boundaries and at given points in time.
Together with the elements, you can choose to query for their tags
and/or their metadata such as the changeset ID, the time of the last
edit or the version number.

The following query extracts the geometries of buildings within 1000 m
of Heidelberg main station with their tags. The response is used to
visualize the buildings and the values of their `building:levels` tag
(if available):

``` r
hd_station_1km <- ohsome_boundary("8.67542,49.40347,1000")

ohsome_elements_geometry(
    boundary = hd_station_1km, 
    filter = "building=* and building!=no and geometry:polygon",
    time = "2021-12-01",
    properties = "tags", 
    clipGeometry = FALSE
) |>
    ohsome_post() |>
    transmute(level = factor(`building:levels`)) |>
    mapview(zcol = "level", lwd = 0, layer.name = "Building level")
```

<img src="man/figures/README-building_levels-1.png" width="900" />

Similarly, you can use `ohsome_elements_centroid()` to extract centroids
of OSM elements and `ohsome_elements_bbox()` for their bounding boxes.
Note that OSM node elements (with point geometries) are omitted from the
results if querying for bounding boxes.

#### Extracting the full history of OSM elements

While the elements extraction endpoints provide geometries and
properties of OSM elements at specific timestamps, results of queries to
the <a
href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html#elements-full-history-extraction"
target="blank">full history endpoints</a> will include all changes to
matching OSM features with corresponding `validFrom` and `validTo`
timestamps.

Here, we request the full history of OSM buildings within 1000 m of
Heidelberg main station, filter for features that still exist and
visualize all building features with their year of creation:

``` r
meta <- ohsome_get_metadata()
start <- meta$extractRegion$temporalExtent[1]
end <- "2023-01-01"

ohsome_elementsFullHistory_geometry(
    boundary = hd_station_1km,
    time = c(start, end),
    filter = "building=* and building!=no and geometry:polygon",
    clipGeometry = FALSE,
    properties = "metadata"
) |> 
    ohsome_post() |>
    janitor::clean_names() |>
    group_by(osm_id) |>
    mutate(year = min(format(valid_from, "%Y"))) |>
    filter(valid_to == lubridate::ymd(end)) |>
    mapview(zcol = "year", lwd = 0, layer.name = "Year of Feature Creation")
```

You may find using `clean_names()` from the
<a href="https://github.com/sfirke/janitor" target="blank">janitor</a>
package helpful in order to remove special characters from column names
in the parsed ohsome API response – just as in the example above.

### OSM contributions

#### Aggregating OSM contributions

With queries to the ohsome API’s <a
href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html#contributions-aggregation"
target="blank">contributions aggregation endpoints</a>, you can get
counts of the contributions provided by users to OSM. The following code
requests the number of deletions of man-made objects at the location of
the hypothetical
<a href="https://en.wikipedia.org/wiki/Null_Island" target="blank">Null
Island</a> per year between 2010 and 2020:

``` r
ohsome_contributions_count(
    boundary = "0,0,10", 
    filter = "man_made=*", 
    time = "2010/2020/P1Y",
    contributionType = "deletion"
) |>
    ohsome_post()
#>    fromTimestamp toTimestamp value
#> 1     2010-01-01  2011-01-01     0
#> 2     2011-01-01  2012-01-01     1
#> 3     2012-01-01  2013-01-01     0
#> 4     2013-01-01  2014-01-01     0
#> 5     2014-01-01  2015-01-01     0
#> 6     2015-01-01  2016-01-01     3
#> 7     2016-01-01  2017-01-01     1
#> 8     2017-01-01  2018-01-01     8
#> 9     2018-01-01  2019-01-01     4
#> 10    2019-01-01  2020-01-01     4
```

The `contributionType` parameter is used to filter for specific types of
contributions (in this case: deletions). If it is not set, any
contribution is counted. Note that the resulting values apply to time
intervals defined by a `fromTimestamp` and a `toTimestamp`.

#### Extracting OSM contributions

The <a
href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html#contributions-extraction"
target="blank">contributions extraction</a> endpoints of the ohsome API
can be used to extract feature geometries of contributions.

In the following example, we extract the centroids of all amenities in
the Berlin city district of Neukölln that have had contributions in
March 2020. Consequently, we filter for features that have had tags
changed and visualize their locations:

``` r
nominatimlite::geo_lite_sf("Berlin Neukoelln", points_only = FALSE) |>
    ohsome_contributions_centroid() |>
    set_filter("amenity=*") |>
    set_time("2020-03,2020-04") |>
    set_properties("contributionTypes") |> 
    ohsome_post() |>
    filter(`@tagChange`) |>
    mapview(layer.name = "Amenities with Tag Changes")
#> Warning: 60 element(s) with empty geometries omitted.
```

<img src="man/figures/README-contribution_extraction-1.png" width="900" />

### OSM users

You can get statistics on the number of users editing specific features
through the <a
href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html#users-aggregation"
target="blank">users aggregation</a> endpoints of the ohsome API.

Here, we show the number of users editing buildings before, during and
after the Nepal earthquake 2015:

``` r
ohsome_users_count(
    boundary = "82.3055,6.7576,87.4663,28.7025",
    filter = "building=* and building!=no and geometry:polygon",
    time = "2015-03-01/2015-08-01/P1M"
) |>
    ohsome_post()
#>   fromTimestamp toTimestamp value
#> 1    2015-03-01  2015-04-01    97
#> 2    2015-04-01  2015-05-01  3489
#> 3    2015-05-01  2015-06-01  3103
#> 4    2015-06-01  2015-07-01   478
#> 5    2015-07-01  2015-08-01   185
```

### Bounding geometries

The ohsome API requires bounding geometries either as bounding polygons
(`bpolys`), bounding boxes (`bboxes`) or bounding circles (`bcircles`)
parameters to the query in a textual form (see
<a href="https://docs.ohsome.org/ohsome-api/stable/boundaries.html"
target="blank">ohsome API documentation</a>). The ohsome R package uses
the generic function `ohsome_boundary()` under the hood to make your
life easier. It accepts a wider range of input geometry formats, while
guessing the right type of bounding geometry.

As seen above, `sf` objects can be passed into the `boundary` argument
of `ohsome_query()` and any of its wrapper functions. You can also
update queries with `set_boundary()`. The `sf` object will be converted
into GeoJSON and passed into the `bpolys` parameter of the query.

If you wish to aggregate or extract OSM elements on administrative
boundaries in the `sf` format, you might want to check out packages such
as <a href="https://github.com/ropensci/rnaturalearth"
target="blank">rnaturalearth</a>,
<a href="https://github.com/rspatial/geodata" target="blank">geodata</a>,
<a href="https://github.com/rspatial/raster" target="blank">raster</a>
(in particular its `getData()` function),
<a href="https://gitlab.com/dickoa/rgeoboundaries"
target="blank">rgeoboundaries</a> or
<a href="https://github.com/dieghernan/nominatimlite"
target="blank">nominatimlite</a> for the acquisition of boundary data
that can be used with `ohsome_boundary()`.

There are also the following methods of `ohsome_boundary()` for other
classes of input geometry objects:

1.  `bbox` objects created with `st_bbox` are converted into a textual
    `bboxes` parameter to the query:

``` r
q <- ohsome_query("users/count") |>
    set_boundary(sf::st_bbox(franconia))

q$body$bboxes
#> [1] "8.97592600000002,48.862505,12.2753535,50.5642245"
```

2.  `matrix` objects created with `sp::bbox()`, `raster::bbox()` or
    `terra::bbox()` are also converted into a textual `bboxes`
    parameter. This even applies for matrices created with
    `osmdata::getbb()` and `tmaptools::bb()`, so that you can
    comfortably acquire bounding boxes from the Nominatim API:

``` r
osmdata::getbb("Kigali") |> 
    ohsome_elements_length(time = "2018/2018-12/P1M", filter = "route=bus") |>
    ohsome_post()
#>     timestamp     value
#> 1  2018-01-01  28251.24
#> 2  2018-02-01  28251.24
#> 3  2018-03-01  29103.11
#> 4  2018-04-01 187174.31
#> 5  2018-05-01 378714.14
#> 6  2018-06-01 473593.82
#> 7  2018-07-01 616330.57
#> 8  2018-08-01 648913.79
#> 9  2018-09-01 754147.47
#> 10 2018-10-01 771767.76
#> 11 2018-11-01 848842.98
#> 12 2018-12-01 860391.88
```

3.  You can pass any `character` object with text in the
    <a href="https://docs.ohsome.org/ohsome-api/stable/boundaries.html"
    target="blank">format allowed by the ohsome API</a> to
    `ohsome_boundary()` – even GeoJSON FeatureCollections. It will
    automatically detect whether you have passed the definition of
    `bpolys`, `bboxes` or `bcircles`. It is possible to use `character`
    vectors where each element represents one geometry:

``` r
c("Circle 1:8.6528,49.3683,1000", "Circle 2:8.7294,49.4376,1000") |>
    ohsome_elements_count(filter = "amenity=*", grouping = "boundary", time = 2021) |>
    ohsome_post()
#> Simple feature collection with 2 features and 3 fields
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: 8.639026 ymin: 49.35931 xmax: 8.743193 ymax: 49.44659
#> Geodetic CRS:  WGS 84
#>   groupByBoundaryId value  timestamp                       geometry
#> 1          Circle 1    16 2021-01-01 POLYGON ((8.666574 49.36834...
#> 2          Circle 2    20 2021-01-01 POLYGON ((8.743193 49.43763...
```

⚠️ While `sf` and `bbox` objects will be automatically transformed to
WGS 84 if in a different coordinate reference system, coordinates in
`character` and `matrix` objects always need to be provided as WGS 84.

### Modifying queries

As seen above, existing `ohsome_query` objects can be modified by
`set_endpoint()`, `set_grouping()`, `set_boundary()` or `set_time()`.
The latter and other functions such as `set_filter()` are just wrappers
around the more generic `set_parameters()`. This can be used to modify
the parameters of a query in any possible way:

``` r
q <- ohsome_elements_count("8.5992,49.3567,8.7499,49.4371")

building_filter <- "building=* and building!=no and geometry:polygon"

q |>
    set_endpoint("ratio", append = TRUE) |>
    set_parameters(
        filter = building_filter,
        filter2 = paste(building_filter, "and building:levels=*"),
        time = "2010/2020/P2Y"
    ) |>
    ohsome_post()
#>    timestamp value value2    ratio
#> 1 2010-01-01   529      3 0.005671
#> 2 2012-01-01 10558      6 0.000568
#> 3 2014-01-01 20962     76 0.003626
#> 4 2016-01-01 25770    739 0.028677
#> 5 2018-01-01 29364   1217 0.041445
#> 6 2020-01-01 31470   1451 0.046107
```

### Grouping

<a href="https://docs.ohsome.org/ohsome-api/v1/group-by.html"
target="blank">Grouping endpoints</a> are available for aggregation
resources and can be used to compute the aggregated results grouped by:

- boundary,
- key,
- tag, and
- type.

In many cases, a grouping by `boundary` can be combined with a grouping
by `tag`. Some of the grouping endpoints require additional query
parameters, e.g. `tag` groupings require a `groupByKey` parameter. Not
all grouping endpoints are available for all aggregation resources –
please consult the
<a href="https://docs.ohsome.org/ohsome-api/v1/group-by.html"
target="blank">ohsome API documentation</a> for details.

You can set the `grouping` argument to any aggregation endpoint wrapper
function (e.g. `ohsome_elements_count(grouping = c("boundary", "tag"))`)
or use `set_grouping()` to modify existing query objects.

### Density and ratio requests

Many <a href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html"
target="blank">aggregation resources</a> have endpoints for requesting
density (i.e. count, length, perimeter or area of features **per area**)
or ratios (of OSM elements satisfying a `filter2` to elements satisfying
a `filter`) instead of or in addition to absolute values.

You can request density or ratio values by setting the `return_value`
argument to aggregation endpoint wrapper functions (e.g. 
`ohsome_elements_count(return_value = "density")`). Mind that ratio
endpoints require an additional `filter2` query parameter. Please
consult the
<a href="https://docs.ohsome.org/ohsome-api/v1/endpoints.html"
target="blank">ohsome API documentation</a> or print
`names(ohsome_endpoints)` to the console in order to check for the
availability of specific density and ratio endpoints.

### Dealing with complex API responses

The ohsome API allows grouping aggregate values for various timestamps
by boundary and tag at the same time. The parsed content of the response
can be rather complex. In the following case, building feature counts
for the districts of Franconia at two different timestamps are requested
– additionally grouped by the building:levels tag. To avoid lots of
redundant geometries, comma-separated values (instead of GeoJSON) are
explicitly requested as the response format:

``` r
building_levels <- franconia |>
    mutate(id  = NUTS_ID) |>
    ohsome_elements_count(grouping = c("boundary", "tag"), format = "csv") |>
    set_filter("building=* and building!=no and geometry:polygon") |>
    set_time("2015/2020") |>
    set_groupByKey("building:levels") |>
    ohsome_post()

dim(building_levels)
#> [1]    2 1999
```

The query results in a confusing data.frame with 1999 columns and 2
rows. This happens because there is a building count column for each
combination of boundary polygon and number of levels, while the two
requested timestamps are in the rows. Fortunately, there is the
<a href="https://tidyr.tidyverse.org" target="blank">tidyr</a> package
to do its magic and pivot this table into a long format with one value
per row:

``` r
library(tidyr)

building_levels |>
    pivot_longer(-timestamp, names_to = c("id", "levels"), names_sep = "_")
#> # A tibble: 3,996 × 4
#>    timestamp           id    levels            value
#>    <dttm>              <chr> <chr>             <dbl>
#>  1 2015-01-01 00:00:00 DE241 remainder          4307
#>  2 2015-01-01 00:00:00 DE241 building.levels.1  6347
#>  3 2015-01-01 00:00:00 DE241 building.levels.2  6727
#>  4 2015-01-01 00:00:00 DE241 building.levels.3  2787
#>  5 2015-01-01 00:00:00 DE241 building.levels.4   745
#>  6 2015-01-01 00:00:00 DE241 building.levels.5    96
#>  7 2015-01-01 00:00:00 DE241 building.levels.6    38
#>  8 2015-01-01 00:00:00 DE241 building.levels.9     7
#>  9 2015-01-01 00:00:00 DE241 building.levels.0     0
#> 10 2015-01-01 00:00:00 DE241 building.levels.7    32
#> # ℹ 3,986 more rows
```

## How to cite this package

In order to cite this package in publications, please use the citation
information provided through `citation("ohsome")`.

------------------------------------------------------------------------

[^1]: `ohsome_post` fails on queries created with an undefined `time`
    parameter unless strict query validation is disabled with
    `strict = FALSE`.

[^2]: Instead of the new R native pipe `|>` you may choose to use
    `magrittr`’s `%>%`.

[^3]: The order of the elements in the character vector is critical!


NEWS.md
# ohsome (development version)

* Uses consistent ohsome filter for building features throughout documentation

# ohsome 0.2.2

* On attach, the package now issues an informative startup message instead of a
warning if it could not retrieve metadata information from the ohsome API.
* Vignette code chunks do not evaluate on CRAN to avoid errors when internet
resources are temporally unavailable
* Fixed tests to use suggested packages conditionally

# ohsome 0.2.1

* Package CITATION file calls `bibentry()` instead of old-style `citEntry()`.
* Fixed incomplete file URIs in docs.
* Fixed missing `strict = FALSE` in query example in `README`.


# ohsome 0.2.0

* Added a `NEWS.md` file to track changes to the package.
* Added `strict`argument to `ohsome_post`. When set to TRUE (default), an error
is thrown on invalid queries and the request is **not** sent to the API. Queries 
with undefined filter or time parameters are considered as invalid in strict 
mode.
* Changed behavior of `set_properties()`: Removes properties parameter from 
query body by default, accepts *tags* and/or *metadata* and/or 
*contributionTypes* as properties argument (multiple values provided as 
comma-separated character or character vector).
* Changed behavior of `set_time()`, `set_filter()`, `set_groupByKeys()`, 
`set_groupByKey()` and `set_groupByValues`: Return unmodified query object if
parameter argument is missing (e.g. `set_filter(query)`), but remove parameter 
from body if explicitly set to NULL (e.g. `set_filter(query, filter = NULL)`)
* Added `grouping` argument to `ohsome_query()` and `set_grouping()`function. 
Based on `grouping`, the endpoint URL is appended so that aggregations are 
grouped accordingly.
* Added `return_value` argument to `ohsome_aggregate_elements()`. Based on 
`return_value`, the endpoint URL is appended so that either absolute aggregate 
values, densities or ratios are requested from the ohsome API.
* Added `ohsome_query()` wrapper functions `ohsome_extract_elements()`,
`ohsome_elements_bbox`, `ohsome_elements_centroid` and `ohsome_elements_geometry` 
for elements extraction endpoints of ohsome API
* Added `ohsome_query()` wrapper functions `ohsome_extract_elementsFullHistory()`,
`ohsome_elementsFullHistory_bbox()`, `ohsome_elementsFullHistory_centroid()` and 
`ohsome_elementsFullHistory_geometry()` for elementsFullHistory extraction
endpoints of ohsome API
* Added `ohsome_query()` wrapper functions `ohsome_extract_contributions()`,
`ohsome_contributions_bbox()`, `ohsome_contributios_centroid()` and 
`ohsome_contributions_geometry()` for contributions extraction endpoints of ohsome 
API
* Added `ohsome_query()` wrapper functions `ohsome_contributions_count()` for 
contributions aggregation endpoints of ohsome API
* Added `ohsome_query()` wrapper function `ohsome_users_count()` for user 
aggregation endpoints of ohsome API
* `ohsome_medata` and `ohsome_temporalExtent` are assigned on loading (not on
attaching) the package
* `ohsome_metadata$extractRegion$temporalExtent` and `ohsome_temporalExtent` are 
no longer converted to POSIXct, but provided as ISO 8601 strings
* Added endpoint-specific check for missing required parameters to query 
validation
* Added validation of JSON response content
* Updated `README` to reflect new features, added hints on boundary polygon
acquisition through third-party packages
* Added `CITATION` file with reference to OSHDB/ohsome API and technical paper


R/ohsome_boundary.R
#' Create an `ohsome_boundary` object
#'
#' Creates an `ohsome_boundary` object from various classes of input geometries. 
#' The `ohsome_boundary` object is used to set the `bpolys`, `bboxes` or
#' `bcircles` parameter of an `ohsome_query` object.
#'
#' @param boundary Bounding geometries specified by WGS84 coordinates in the 
#'   order `lon,lat`. The geometries of `sf` are transformed to WGS84 if the CRS 
#'   of the object is known. The following classes are supported:
#'   * `sf` with (MULTI)POLYGON geometries
#'   * `sfc` with (MULTI)POLYGON geometries
#'   * `sfg` with (MULTI)POLYGON geometries and WGS 84 coordinates
#'   * `bbox` created with [sf::st_bbox()] or [tmaptools::bb()]
#'   * `matrix` created with [sp::bbox()] or [osmdata::getbb()]
#'   * `character` providing textual definitions of bounding polygons, boxes or 
#'   circles as allowed by the ohsome API (see 
#'   [ohsome API - Boundaries](https://docs.ohsome.org/ohsome-api/stable/boundaries.html)
#'   ):
#'     * bboxes: WGS84 coordinates in the following format: 
#'     `"id1:lon1,lat1,lon2,lat2|id2:lon1,lat1,lon2,lat2|..."` OR 
#'     `"lon1,lat1,lon2,lat2|lon1,lat1,lon2,lat2|..."`
#'     * bcircles: WGS84 coordinates + radius in meter in the following 
#'     format: `"id1:lon,lat,r|id2:lon,lat,r|..."` OR 
#'     `"lon,lat,r|lon,lat,r|..."`
#'     * bpolys: WGS84 coordinates given as a list of coordinate pairs (as for 
#'     bboxes) or GeoJSON FeatureCollection. The first point has to be the same 
#'     as the last point and MultiPolygons are only supported in GeoJSON.
#'   * `list` of `bbox`, `matrix` or `character`. Bounding geometry types of all 
#'   list elements must be the same. Does not work with GeoJSON 
#'   FeatureCollections.
#' @param ... Additional arguments other than `digits` are ignored.
#' @param digits integer; number of decimal places of coordinates in the 
#'   resulting GeoJSON when converting `sf` to GeoJSON (defaults to 6).
#' @return An `ohsome_boundary` object which contains the following elements: 
#'   * `boundary`: the boundary in textual format
#'   * `type` of the boundary (`bpolys`, `bcircles`, or `bboxes`).
#' @export
#' @examples
#' # Defintion of a bounding circle (lon,lat,radius in meters)
#' ohsome_boundary("8.6528,49.3683,1000") 
#' 
#' # Definition of two named bounding circles
#' ohsome_boundary("Circle 1:8.6528,49.3683,1000|Circle 2:8.7294,49.4376,1000")
#' 
#' # Definition of two named bounding circles with a character vector
#' ohsome_boundary(c("Circle 1:8.6528,49.3683,1000", "Circle 2:8.7294,49.4376,1000"))
#' 
#' # Use franconia from the mapview package as bounding polygons
#' \donttest{
#' ohsome_boundary(mapview::franconia, digits = 4)
#' }
#'  
#' # Use the bounding box of franconia
#' \donttest{
#' ohsome_boundary(sf::st_bbox(mapview::franconia))
#' }
#' 
#' # Get bounding box of the city of Berlin from OSM
#' \dontrun{
#' ohsome_boundary(osmdata::getbb("Berlin"))
#' }
#'  
#' # Use a list of two bounding boxes
#' \dontrun{
#' ohsome_boundary(list(osmdata::getbb("Berlin"), sf::st_bbox(mapview::franconia)))
#' }
#' 
ohsome_boundary <- function(boundary, ...) UseMethod("ohsome_boundary")

#' @name ohsome_boundary
#' @export
ohsome_boundary.ohsome_boundary <- function(boundary, ...) return(boundary)

#' @name ohsome_boundary
#' @export
ohsome_boundary.character <- function(boundary, ...) {

	if(jsonlite::validate(boundary) && length(boundary) == 1) {
			type <- "bpolys"
	} else {
		splt <- unlist(strsplit(boundary, split = "|", fixed = TRUE))
		len <- unique(sapply(strsplit(splt, split = ","), length))

		if(length(len) == 1 && len == 3) {
			type <- "bcircles"
		} else if(length(len) == 1 && len == 4) {
			type <- "bboxes"
		} else type <- "bpolys"
	}

	structure(list(
		boundary = paste(boundary, collapse = "|"), type = type),
		class = "ohsome_boundary"
	)
}

#' @name ohsome_boundary
#' @export
ohsome_boundary.sf <- function(boundary, digits = 6, ...) {

	if( (sf::st_crs(boundary)$input != "EPSG:4326") & !is.na(sf::st_crs(boundary)) )
	  boundary <- sf::st_transform(boundary, 4326)

	types <- sf::st_geometry_type(boundary)

	if(!any(types %in% c("POLYGON", "MULTIPOLYGON"))) {
		stop(
			"None of the geometries provided are of type (MULTI)POLYGON.",
			call. = FALSE
		)
	}

	if(!all(types %in% c("POLYGON", "MULTIPOLYGON"))) {
		warning(
			"At least one of the bounding geometries is not of type ",
			"(MULTI)POLYGON and was omitted.",
			call. = FALSE
		)
		boundary <- boundary[types %in% c("POLYGON", "MULTIPOLYGON"),]
	}

	geojson <- geojsonsf::sf_geojson(boundary, digits = digits, simplify = FALSE)
	type <- "bpolys"

	structure(list(boundary = geojson, type = type), class = "ohsome_boundary")
}

#' @name ohsome_boundary
#' @export
ohsome_boundary.sfc <- function(boundary, ...) ohsome_boundary(sf::st_as_sf(boundary), ...)

#' @name ohsome_boundary
#' @export
ohsome_boundary.sfg <- function(boundary, ...) ohsome_boundary(sf::st_sfc(boundary, crs = 4326), ...)

#' @name ohsome_boundary
#' @export
ohsome_boundary.bbox <- function(boundary, ...) {

	if( (attributes(boundary)$crs$input != "EPSG:4326") & (!is.na((attributes(boundary)$crs$input)) ) ) {
		boundary <- sf::st_bbox(sf::st_transform(sf::st_as_sfc(boundary), 4326))
	}

	ohsome_boundary(paste(boundary, collapse = ","))
}

#' @name ohsome_boundary
#' @export
ohsome_boundary.matrix <- function(boundary, ...) ohsome_boundary(paste(boundary, collapse = ","))

#' @name ohsome_boundary
#' @export
ohsome_boundary.list <- function(boundary, ...) {
	ohsome_boundary(paste(sapply(boundary, ohsome_boundary)[1,], collapse = "|"))
}
